src/GafferUI/GraphGadget.cpp

```diff
@@ -223,6 +232,230 @@ class SetPositionsAction : public Gaffer::Action
+					{
+						activeOutput = outputNameValue->valuePlug();
+					}
+
+					// We are handling just the value inputs specially
+					for( auto &inSourcePlug : Gaffer::Plug::InputRange( *nameSwitchNode->inPlugs() ) )
+					{
+						const Gaffer::NameValuePlug *inputNameValue = runTimeCast<const Gaffer::NameValuePlug>( inSourcePlug.get() );
+
+						handledBySpecialCase.push_back( inputNameValue->valuePlug() );
+					}
+				}
+				else
+				{
+					if( connectionStart == switchNode->outPlug() )
+					{
+						activeOutput = connectionStart;
+					}
+
+					for( auto &inSourcePlug : Gaffer::Plug::InputRange( *switchNode->inPlugs() ) )
+					{
+						handledBySpecialCase.push_back( inSourcePlug.get() );
+					}
+				}
+
+				const Gaffer::Plug *active = nullptr;
+				if( activeOutput )
+				{
+					Gaffer::Context::Scope s( context );
+					active = switchNode->activeInPlug( activeOutput );
+				}
+
+				// Track the active input plug
+				if( active )
+				{
+					const Gaffer::Plug *input = active->getInput();
+					if( input )
+					{
+						activeWalk( input, active, context, canceller, activePlugs, activeNodes, plugContextsVisited );
+					}
+				}
+			}
+			else if( const Gaffer::ContextProcessor *contextProcessorNode = runTimeCast<const Gaffer::ContextProcessor>( node ) )
+			{
+				const Gaffer::Plug *input = contextProcessorNode->inPlug()->getInput();
+				if( input )
+				{
+					Gaffer::Context::Scope s( context );
+					const Gaffer::ContextPtr newContext = contextProcessorNode->inPlugContext();
+					activeWalk( input, contextProcessorNode->inPlug(), newContext.get(), canceller, activePlugs, activeNodes, plugContextsVisited );
+				}
+				handledBySpecialCase.push_back( contextProcessorNode->inPlug() );
+			}
+		}
+
+		if( const Gaffer::Loop *loopNode = runTimeCast<const Gaffer::Loop>( node ) )
+		{
+			if(
+				!firstVisit &&
+				( connectionStart == loopNode->previousPlug() ||  connectionStart->parent() == loopNode->previousPlug() )
+			)
+			{
+				// In any normal circumstance, if we've already visited this loop, that means we will
+				// have already done all needed traversal, and just stop here.
+				//
+				// Stopping here avoids leaking out the traversal from inside the loop where we don't
+				// know the context.
+				//
+				// There are some weird cases where this might not be totally accurate - it's technically
+				// possible to make a graph that uses the inner nodes of a loop both through the loop, and
+				// also through manually using a ContextVariables node to set loop:index - this is a very
+				// rare special case though
+				return;
+			}
+
+			// The input plug of the loop is a normal input, and will be handled below
+
+			// The next plug of the loop is evaluated in many different contexts - instead of evaluating all
+			// of them, pass a null context which will trigger a conservative evaluation of all possible inputs
```

Comment:
Yeah, maybe we will want to do that eventually, though I think it would have some knock-on effects - currently, there is almost no reasonable way to make this take a long time, which allows skipping some concerns.  If we add real support for loops, then it possible for it to take a long time, and maybe we would also want to start using multiple threads for evaluation, and add some sort of busy indicator?

Reply:
> currently, there is almost no reasonable way to make this take a long time, which allows skipping some concerns

For some definitions of reasonable :) In practice, I'll wager that people are still merrily doing things like connecting FilterResults to `SomeNode.enabled` somewhere.